aws/data_source_aws_regions_test.go

```diff
@@ -0,0 +1,119 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func TestAccDataSourceAwsRegions_Basic(t *testing.T) {
+	resourceName := "data.aws_regions.empty"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceAwsRegionsConfig_empty(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDataSourceAwsRegions_OptIn(t *testing.T) {
+	resourceName := "data.aws_regions.opt_in_status"
+
+	statusOptedIn := "opted-in"
+	statusNotOptedIn := "not-opted-in"
+	statusOptInNotRequired := "opt-in-not-required"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			// This resource has to be at the very top of the test scenario due to bug in Terrafom Plugin SDK
+			{
+				Config: testAccDataSourceAwsRegionsConfig_allRegionsFiltered(statusOptedIn),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+			{
+				Config: testAccDataSourceAwsRegionsConfig_allRegionsFiltered(statusOptInNotRequired),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+			{
+				Config: testAccDataSourceAwsRegionsConfig_allRegionsFiltered(statusNotOptedIn),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+		},
+	})
+}
```

Comment:
This test appears to be testing the API more than the Terraform data source itself -- we can likely pare this down to a single filter case that should always pass:

```suggestion
func TestAccDataSourceAwsRegions_Filter(t *testing.T) {
	resourceName := "data.aws_regions.opt_in_status"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:  func() { testAccPreCheck(t) },
		Providers: testAccProviders,
		Steps: []resource.TestStep{
			{
				Config: testAccDataSourceAwsRegionsConfig_allRegionsFiltered("opt-in-not-required"),
				Check: resource.ComposeTestCheckFunc(
					testAccDataSourceAwsRegionsCheck(resourceName),
				),
			},
		},
	})
}
```